File: app/src/main/java/com/ed/redditapp/ui/main/MainSubredditIconUrlAsyncTask.java
package com.ed.redditapp.ui.main;

import com.ed.redditapp.lib.AsyncTask;
import com.ed.redditapp.lib.api.Post;
import com.ed.redditapp.lib.api.RedditApi;

public class MainSubredditIconUrlAsyncTask {
    interface Callback {
        void callback(String iconUrl, int position);
    }

    private final RedditApi redditApi;

    public MainSubredditIconUrlAsyncTask(RedditApi redditApi) {
        this.redditApi = redditApi;
    }

    public void run(Post[] posts, Callback callback) {
        for (int i = 0; i < posts.length; i++) {
            int[] index = {i};
            Post post = posts[i];
            new AsyncTask().runAsync((h) -> {
                String iconUrl = redditApi.getSubredditIconUrl(post.getSubreddit());
                h.onUIThread(() -> callback.callback(iconUrl, index[0]));
            });
        }
    }
}

File: app/src/main/java/com/ed/redditapp/ui/main/MainSubredditIconUrlAsyncTask.kt
package com.ed.redditapp.ui.main

import com.ed.redditapp.lib.AsyncTask
import com.ed.redditapp.lib.api.Post
import com.ed.redditapp.lib.api.RedditApi

typealias MainSubredditIconUrlCallback = (iconUrl: String, position: Int) -> Unit

class MainSubredditIconUrlAsyncTask(private val redditApi: RedditApi) {
    fun run(posts: Array<Post>, callback: MainSubredditIconUrlCallback) {
        for (i in IntRange(0, posts.size)) {
            AsyncTask().runAsync { h ->
                val iconUrl = redditApi.getSubredditIconUrl(posts[i].subreddit)
                h.onUIThread { callback(iconUrl, i) }
            }
        }
    }
}